### PR TITLE
feat(diagnostics): configure publish interval

### LINKED ISSUE #793: config: make the publishDiagnostics quiet window configurable (minPublishIntervalMs)
## Context

PR #685 added a per-host **quiet window** on the editor-facing
`textDocument/publishDiagnostics` wire sends to stop the diagnostics wire flood:
a burst of merged-set changes during a typing storm collapses into ≤1 full-set
publish per host per window. The window is currently a hardcoded constant.

`src/lsp/lsp_impl/coordinator/diagnostic_publisher.rs:90`
```rust
const WIRE_PUBLISH_QUIET_WINDOW: std::time::Duration = std::time::Duration::from_secs(1);
```

It is a **leading-edge throttle**: the first publish after quiet passes through
immediately (no latency outside bursts); a changed merge inside the window is
withheld and exactly one trailing task re-publishes the *latest* merge at the
window's end. It is keyed **per host**
(`wire_gate_admit(host, WIRE_PUBLISH_QUIET_WINDOW)` at `:725`).

1 s is a reasonable default, but it is a one-size-fits-all transport-fairness
knob. Hosts with light diagnostic traffic gain nothing from coalescing (it only
adds latency), and users who prioritize freshness over stdout fairness have no
way to shorten it. Make it configurable.

## Proposal

Add a per-method field on the cross-layer aggregation block, co-located with the
existing wire seal (`priorities = []`):

```toml
[languages.markdown.layers.aggregation."textDocument/publishDiagnostics"]
priorities = []              # existing: wire seal (do not publish at all)
minPublishIntervalMs = 1000  # new: min interval between publish wire sends
                             #      (0 = publish on every change)
```

Full path:
`languages.<host-lang>.layers.aggregation."textDocument/publishDiagnostics".minPublishIntervalMs`

- Type: `Option<u64>` (milliseconds). `None` → current default (`1000`).
  `0` → disable coalescing (leading pass-through only; publish on every change).
- Field on `LayerAggregationConfig` (`src/config/settings.rs:152`), sibling to
  `priorities` / `strategy`. `#[serde(rename_all = "camelCase")]` →
  TOML key `minPublishIntervalMs`.

### Why `layers.aggregation`, not `bridge.<lang2>.aggregation`

The wire gate is **one per host URI**, not per injected language. A single host
merges diagnostics from many `bridge.<lang2>` servers (python + lua + ts) into
one wire send. A per-injected-language config cannot map onto a per-host gate
(which `<lang2>`'s window would win?). `layers.aggregation` is keyed by
host-language × method — the exact granularity of the gate — and is where the
sibling seal already lives.

### Why `minPublishInterval`, not `minUpdateInterval`

The window throttles the **push / publish** channel only. The pull path
(`textDocument/diagnostic`) is a separate mechanism (PR #685's content-addressed
`resultId` / `unchanged` reports) that this window does **not** touch.
`Update` would over-promise on pull-first setups (nvim ≥ 0.11) where diagnostics
arrive via pull and the window has no effect. `Publish` scopes it correctly and
mirrors the parent method key `publishDiagnostics` rather than introducing new
vocabulary.

## Implementation sketch

1. Add `min_publish_interval_ms: Option<u64>` to `LayerAggregationConfig`
   (`src/config/settings.rs:152`).
2. Carry it on `ResolvedLayerConfig` and resolve via the existing
   `resolve_layer_config_from_settings(settings, host_lang, "textDocument/publishDiagnostics")`
   path — the same resolver `publish_sealed()` already uses
   (`diagnostic_publisher.rs:819`).
3. In `republish` (`diagnostic_publisher.rs:725`), pass the resolved `Duration`
   into `wire_gate_admit` instead of the constant. `WIRE_PUBLISH_QUIET_WINDOW`
   becomes the default when unset.
4. `0` → skip the gate (send immediately on every change).

Plumbing is light: `wire_gate_admit` already takes the window as an argument.

## Scope / non-goals

- **Push (publishDiagnostics) wire only.** Does not affect pull cadence, the
  seal, change detection, or `workspace/diagnostic/refresh` nudging — those fire
  at the original points regardless.
- Per-host-language granularity via the existing method-keyed block; not
  per-injected-language, not global-only.
- Docs: the schema doc should describe the **user-visible effect** ("during
  rapid typing, diagnostics are published at most once per this interval;
  `0` = publish on every change"), not the internal wire/throttle vocabulary.

## References

- PR #685 (introduced the hardcoded window)
- `src/lsp/lsp_impl/coordinator/diagnostic_publisher.rs:90,725,819`
- `src/config/settings.rs:152` (`LayerAggregationConfig`)

### PR DESCRIPTION
**Pull request by atusy**
Closes #793

## Summary

- add per-host-language minPublishIntervalMs resolution for textDocument/publishDiagnostics
- preserve the 1000 ms default, allow 0 for immediate publishing, and cap extreme values at 24 hours
- make pending publishes react to live configuration changes without losing cancellation debt
- re-evaluate open hosts after configuration reloads, including sealed-to-unsealed transitions
- document the setting and aggregation contract

## Validation

- cargo fmt --check
- cargo test --lib -q zero_window_uncommitted_send_retries
- cargo test --lib -q live_zero_interval_flushes_existing_withheld_publish
- cargo test --lib -q zero_publish_interval_never_withholds_changed_merges
- cargo test --lib -q trailing_republish_flushes_the_withheld_wire_send
- cargo test --lib -q settings_change_republishes_an_unchanged_set_after_unsealing
- cargo test --lib -q republish_under_seal_keeps_the_change_contract
- cargo clippy --lib -- -D warnings
- git diff --check

## Review

- multi-perspective review: converged
- Codex iterative review: converged
- fresh Codex review: no actionable findings

This PR intentionally remains Draft until the full revise pipeline finishes.

## Summary by CodeRabbit

- **New Features**
  - Added `minPublishIntervalMs` to throttle proactive, editor-facing `textDocument/publishDiagnostics` per language/method, defaulting to `1000` and capped at 24 hours; set to `0` to publish immediately.
- **Bug Fixes**
  - Improved republishing after settings reloads, including contract-change detection and timely flushing after unsealing.
  - Fixed edge-case behavior for a zero-duration quiet window.
- **Documentation**
  - Updated configuration docs and architecture notes with the new option, defaults, limits, and scope.
- **Tests**
  - Expanded coverage for quiet-window throttling, settings reload republish flow, and JSON deserialization.

### CODE REVIEW

**Comment by atusy** [comment]
@copilot review

**Comment by atusy** [comment]
@coderabbitai review

**Comment by atusy** [comment]
@coderabbitai review

File: src/lsp/lsp_impl/coordinator/diagnostic_publisher.rs [hunk 1]
```diff
@@ -755,15 +757,47 @@ impl DiagnosticPublisher {
...
+            &language_name,
+            "textDocument/publishDiagnostics",
+        );
+        let current = crate::lsp::lsp_impl::bridge_context::resolve_layer_config_from_settings(
+            &self.settings_manager.load_settings(),
+            &language_name,
+            "textDocument/publishDiagnostics",
+        );
+        previous != current
+    }
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/lsp/lsp_impl/coordinator/diagnostic_publisher.rs:889]: Fixed in 9654eecfa: contract comparison now normalizes None to the 1s default and caps intervals at 24h before comparing, with regressions for both equivalence classes.

File: src/lsp/lsp_impl/coordinator/diagnostic_publisher.rs [hunk 2]
```diff
@@ -755,15 +757,47 @@ impl DiagnosticPublisher {
...
+    /// Whether replacing `previous` with the currently published settings
+    /// changes this host's editor-facing `publishDiagnostics` contract.
+    pub(crate) fn publish_contract_changed(
+        &self,
+        host: &Url,
+        previous: &crate::config::WorkspaceSettings,
+    ) -> bool {
+        let Some(language_name) = self.open_document_language(host) else {
+            return true;
+        };
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/lsp/lsp_impl/coordinator/diagnostic_publisher.rs]: Clarified in 9654eecfa: unknown-language hosts are conservatively treated as changed to avoid retaining stale wire state.

File: src/lsp/lsp_impl/coordinator/diagnostic_publisher.rs [hunk 3]
```diff
@@ -720,10 +721,11 @@ impl DiagnosticPublisher {
...
         // coverage bump + refresh nudge) is unaffected.
-        match self
-            .aggregator
-            .wire_gate_admit(host, WIRE_PUBLISH_QUIET_WINDOW)
-        {
+        // Subscribe before resolving/admitting so a settings replacement in
+        // that window remains observable by a newly parked trailing task.
+        let settings_changed = self.settings_manager.subscribe_settings_changes();
+        let quiet_window = self.publish_quiet_window(host);
+        match self.aggregator.wire_gate_admit(host, quiet_window) {
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/lsp/lsp_impl/coordinator/diagnostic_publisher.rs]: Not changed: subscribing before wire_gate_admit is intentional. Moving subscription into schedule_trailing creates a race where settings change after admit but before subscribe, leaving the newly parked task asleep on an already-published watch value. The receiver allocation is the cost of closing that correctness window.

File: src/lsp/lsp_impl/coordinator/diagnostic_publisher.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (atusy) [reply src/lsp/lsp_impl/coordinator/diagnostic_publisher.rs:889]: Fixed in `9654eecfa`. Contract comparison now normalizes both settings snapshots to `(sealed, effective_quiet_window)`, including the default and 24-hour cap, with discriminating tests for `None == 1000ms` and oversized values.

File: src/lsp/lsp_impl/coordinator/diagnostic_publisher.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (atusy) [reply src/lsp/lsp_impl/coordinator/diagnostic_publisher.rs]: Documented in `9654eecfa`: unknown languages conservatively invalidate prior publish state rather than retaining a potentially stale wire contract.

File: src/lsp/lsp_impl/coordinator/diagnostic_publisher.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (atusy) [reply src/lsp/lsp_impl/coordinator/diagnostic_publisher.rs]: Refuted: the subscription must be created before both quiet-window resolution and gate admission. If it were moved inside `schedule_trailing`, a settings replacement between reading the old quiet window and subscribing could be missed; the newly parked task would then sleep the stale interval with no watch notification. The receiver allocation on SendNow/already-scheduled paths is the cost of closing that lost-wakeup window, as the existing lines 724-725 comment states.

**Comment by atusy** [comment]
Closing in favor of #851. The replacement uses a global method policy with per-URI leading + trailing debounce (`debounceMs` / `maxWaitMs`) instead of a per-language minimum publish interval.
